mac/Focalboard/ViewController.swift

```diff
@@ -64,20 +78,70 @@ class ViewController:
 	@objc func onServerStarted() {
 		NSLog("onServerStarted")
 		DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
-			self.updateSessionToken()
+			self.updateSessionTokenAndUserSettings()
 			self.loadHomepage()
 		}
 	}
 
-	private func updateSessionToken() {
+	private func persistUserSettings() {
+		let semaphore = DispatchSemaphore(value: 0)
+
+		// Convert to base64 to avoid escaping issues when later inserting the exported settings into the import script
+		webView.evaluateJavaScript("window.btoa(Focalboard.exportUserSettings())") { result, error in
+			defer { semaphore.signal() }
+			guard let base64 = result as? String, let data = Data(base64Encoded: base64), let decoded = String(data: data, encoding: .utf8) else {
+				NSLog("Failed to export user settings: \(error?.localizedDescription ?? "?")")
+				return
+			}
+			UserDefaults.standard.set(base64, forKey: "localStorage")
+			NSLog("Persisted user settings: \(decoded)")
+		}
+
+		// During shutdown the system grants us about 5 seconds to clean up and store user data
+		let timeout = DispatchTime.now() + .seconds(3)
+		var result: DispatchTimeoutResult?
+
+		// Busy wait because evaluateJavaScript can only be called from *and* signals on the main thread
+		while (result != .success && .now() < timeout) {
+			result = semaphore.wait(timeout: .now())
+			RunLoop.current.run(mode: .default, before: Date())
+		}
+
+		if result == .timedOut {
+			NSLog("Timed out trying to persist user settings")
+		}
+	}
+
+	private func updateSessionTokenAndUserSettings() {
 		let appDelegate = NSApplication.shared.delegate as! AppDelegate
-		let script = WKUserScript(
+		let sessionTokenScript = WKUserScript(
 			source: "localStorage.setItem('focalboardSessionId', '\(appDelegate.sessionToken)');",
 			injectionTime: .atDocumentStart,
 			forMainFrameOnly: true
 		)
+		let base64 = UserDefaults.standard.string(forKey: "localStorage") ?? ""
+		let userSettingsScript = WKUserScript(
+			source: """
+				const settings = window.atob("\(base64)");
```

Comment:
It would be nice if the btoa / atob functions could be pushed to the shared TS code, e.g. `importUserSettingsBlob`, `exportUserSettingsBlob`, that could just wrap the existing methods. Minor, but it would mean less brittle code in the native apps.

Reply:
Good call. Made the change.